crates/shared/src/gas_price_estimation.rs

```diff
@@ -128,3 +132,76 @@ impl GasPriceEstimating for FakeGasPriceEstimator {
         Ok(*self.0.lock().unwrap())
     }
 }
+
+/// Gas estimator which will never reuse an old gas price estimate. This is useful to supply a
+/// cheap finger print for otherwise identical transactions.
+/// In case an estimate would be a duplicate the estimator increases the estimate until it finds
+/// the first unique value.
+/// To keep the unique prices as close to the expected prices as possible a
+/// `FingerPrintingGasEstimator` should be as short lived as possible.
+pub struct FingerPrintingGasPriceEstimator<'a> {
+    inner: &'a dyn GasPriceEstimating,
+    previous_estimates: Mutex<HashSet<u64>>,
+}
+
+impl<'a> FingerPrintingGasPriceEstimator<'a> {
+    pub fn new(inner: &'a dyn GasPriceEstimating) -> Self {
+        Self {
+            inner,
+            previous_estimates: Default::default(),
+        }
+    }
+
+    fn closest_unique_gas_estimate(&self, mut target: EstimatedGasPrice) -> EstimatedGasPrice {
+        let mut closest_unique_price = target.legacy;
+        let mut previous_estimates = self.previous_estimates.lock().unwrap();
+        while !previous_estimates.insert(closest_unique_price.to_bits()) {
```

Comment:
So, here we are filling the `previous_estimates`. Is there a place where we occasionally clear this variable? Or this is only done on app restart?

Reply:
(As I understand we need unique hashes across currently submitted and pending transactions, not across all transactions we ever submitted?)

Reply:
The idea was to manage `previous_estimates` by creating `FingerPrintingGasPriceEstimator` which only live as long as we need unique prices. So every settlement we try to bring on chain would get its own `FingerPrintingGasPriceEstimator` but it seems this whole stuff is not required at all. As mentioned in my other comment 👇